### DISCUSSION TITLE
c++ macros seem to complete differently

### DISCUSSION THREAD

**Discussion by ville-h** [company-mode/company-mode#1457] [Q&A]
Whilst programming in c++ I noticed macros seem to complete differently to other things. Once again I am not sure if this is my setup or if this is a company bug so I'll start as discussion rather than bug report.

My `company-backends` is `((company-capf company-yasnippet))`. My `completion-styles` is `(basic initials partial-completion substring)`. I am using eglot and yasnippet. Here is a simple test case to demonstrate:

```
#define ABC_012


#define def_012


void GHI_012() {
}


void jkl_012() {
}


int main() {
    // 2024-02-19 note: typing "A0" below doesn't complete to "ABC_012", typing "A" does

    // 2024-02-19 note: typing "d0" below doesn't complete to "def_012", typing "d" does

    // 2024-02-19 note: typing "G0" below completes to "GHI_012"

    // 2024-02-19 note: typing "j0" below completes to "jkl_012"

    return 0;
}
```

**Comment by dgutov** [comment]
If you're using Eglot (and not `company-clang`, which is at least part of this package), then the framework cannot know the difference between a function and a macro, right?

This seems like something you could bright up with the Eglot devs. The behavior could also differ between language servers (e.g. `ccls` and `clangd`).

**Comment by ville-h** [comment]
Looks like eglot comes with `eglot-events-buffer` and based on it I'm fairly alright to call it a clangd bug. Below is the output after inserting the "0" in "A0":
```
(:jsonrpc "2.0" :method "textDocument/didChange" :params
    (:textDocument
        (:uri "file:///home/ville/projects/test/test013/src/main.cxx" :version 3)
        :contentChanges
        [(:range
             (:start
                 (:line 16 :character 5)
                 :end
                 (:line 16 :character 5))
             :rangeLength 0 :text "0")]))
[client-request] (id:15) Wed Feb 21 00:05:04 2024:
(:jsonrpc "2.0" :id 15 :method "textDocument/completion" :params
    (:textDocument
        (:uri "file:///home/ville/projects/test/test013/src/main.cxx")
        :position
        (:line 16 :character 6)
        :context
        (:triggerKind 1)))
[server-reply] (id:15) Wed Feb 21 00:05:04 2024:
(:id 15 :jsonrpc "2.0" :result
    (:isIncomplete t :items
        []))
```
The items array at the end should hold the completion candidates I believe, based off of what is returned for just writing "A".

**Reply by dgutov** [reply]
That makes sense.

You can try `lsp-mode` too (sometimes the client could contain workarounds for specific problems), or a different language server.
